Libraries/LibJS/Runtime/Function.h

```diff
@@ -40,12 +40,28 @@ class Function : public Object {
     virtual const FlyString& name() const = 0;
     virtual LexicalEnvironment* create_environment() = 0;
 
+    BoundFunction* bind(Value this_value, Vector<Value> arguments);
+
+    Optional<Value> bound_this() const
+    {
+        return m_bound_this;
+    }
+
+    const Vector<Value>& bound_arguments() const
+    {
+        return m_bound_arguments;
+    }
+
 protected:
     explicit Function(Object& prototype);
     virtual const char* class_name() const override { return "Function"; }
 
+    void bind_this_and_arguments(Value bound_this, Vector<Value> bound_arguments);
+
 private:
     virtual bool is_function() const final { return true; }
+    Optional<Value> m_bound_this;
+    Vector<Value> m_bound_arguments;
```

Comment:
To ensure that these don't get garbage collected, we have to mark them in `Function::visit_children()`.
See for example how `ScriptFunction` marks its parent lexical environment.